File: python/datetime_operations.py
"""
Date and time utility functions for test automation
"""

import time
from datetime import datetime


def get_today_date(date_format: str = "%Y-%m-%d") -> str:
    """
    Get today's date in the specified format
    
    Args:
        date_format: Date format pattern (e.g., "%Y-%m-%d", "%m/%d/%Y", "%d-%b-%Y")
        
    Returns:
        Formatted date string
    """
    return datetime.now().strftime(date_format)


def get_current_timestamp(timestamp_format: str = "%Y-%m-%d %H:%M:%S") -> str:
    """
    Get current timestamp in the specified format
    
    Args:
        timestamp_format: Timestamp format pattern (e.g., "%Y-%m-%d %H:%M:%S", "%Y%m%d_%H%M%S")
        
    Returns:
        Formatted timestamp string
    """
    return datetime.now().strftime(timestamp_format)


def wait_for(seconds: float) -> None:
    """
    Wait for specified seconds
    
    Args:
        seconds: Time to wait in seconds
    """
    time.sleep(seconds)


def wait_for_duration(duration: str) -> None:
    """
    Wait for a duration specified in a human-readable format
    Supports: d (days), h (hours), m (minutes), s (seconds)
    Examples: "2d5h10m", "30m", "1h30m", "2d", "45s"
    
    Args:
        duration: Duration string (e.g., "2d5h10m")
    """
    total_seconds = _parse_duration(duration)
    time.sleep(total_seconds)


def _parse_duration(duration: str) -> float:
    """
    Parse a duration string into seconds
    
    Args:
        duration: Duration string (e.g., "2d5h10m30s")
        
    Returns:
        Total seconds
    """
    total_seconds = 0.0
    number_buffer = ""
    
    for char in duration:
        if char.isdigit():
            number_buffer += char
        else:
            if number_buffer:
                value = int(number_buffer)
                
                char_lower = char.lower()
                if char_lower == 'd':
                    total_seconds += value * 24 * 60 * 60
                elif char_lower == 'h':
                    total_seconds += value * 60 * 60
                elif char_lower == 'm':
                    total_seconds += value * 60
                elif char_lower == 's':
                    total_seconds += value
                else:
                    raise ValueError(f"Invalid duration unit: {char}")
                
                number_buffer = ""
    
    return total_seconds

File: dotNet/DateTimeOperations.cs
using System;
using System.Threading;

namespace TestAutomation.Utils
{
    /// <summary>
    /// Date and time utility functions for test automation
    /// </summary>
    public static class DateTimeOperations
    {
        /// <summary>
        /// Get today's date in the specified format
        /// </summary>
        /// <param name="format">Date format pattern (e.g., "yyyy-MM-dd", "MM/dd/yyyy", "dd-MMM-yyyy")</param>
        /// <returns>Formatted date string</returns>
        public static string GetTodayDate(string format)
        {
            return DateTime.Now.ToString(format);
        }

        /// <summary>
        /// Get current timestamp in the specified format
        /// </summary>
        /// <param name="format">Timestamp format pattern (e.g., "yyyy-MM-dd HH:mm:ss", "yyyyMMdd_HHmmss")</param>
        /// <returns>Formatted timestamp string</returns>
        public static string GetCurrentTimestamp(string format)
        {
            return DateTime.Now.ToString(format);
        }

        /// <summary>
        /// Wait for specified milliseconds
        /// </summary>
        /// <param name="milliseconds">Time to wait in milliseconds</param>
        public static void WaitFor(int milliseconds)
        {
            Thread.Sleep(milliseconds);
        }

        /// <summary>
        /// Wait for a duration specified in a human-readable format
        /// Supports: d (days), h (hours), m (minutes), s (seconds)
        /// Examples: "2d5h10m", "30m", "1h30m", "2d", "45s"
        /// </summary>
        /// <param name="duration">Duration string (e.g., "2d5h10m")</param>
        public static void WaitForDuration(string duration)
        {
            long totalMilliseconds = ParseDuration(duration);
            Thread.Sleep((int)totalMilliseconds);
        }

        /// <summary>
        /// Parse a duration string into milliseconds
        /// </summary>
        /// <param name="duration">Duration string (e.g., "2d5h10m30s")</param>
        /// <returns>Total milliseconds</returns>
        private static long ParseDuration(string duration)
        {
            long totalMilliseconds = 0;
            var numberBuffer = new System.Text.StringBuilder();

            foreach (char c in duration)
            {
                if (char.IsDigit(c))
                {
                    numberBuffer.Append(c);
                }
                else
                {
                    if (numberBuffer.Length > 0)
                    {
                        long value = long.Parse(numberBuffer.ToString());

                        switch (char.ToLower(c))
                        {
                            case 'd':
                                totalMilliseconds += value * 24 * 60 * 60 * 1000;
                                break;
                            case 'h':
                                totalMilliseconds += value * 60 * 60 * 1000;
                                break;
                            case 'm':
                                totalMilliseconds += value * 60 * 1000;
                                break;
                            case 's':
                                totalMilliseconds += value * 1000;
                                break;
                            default:
                                throw new ArgumentException($"Invalid duration unit: {c}");
                        }

                        numberBuffer.Clear();
                    }
                }
            }

            return totalMilliseconds;
        }
    }
}
